Extension/src/LanguageServer/client.ts

```diff
@@ -1043,6 +1052,61 @@ export class DefaultClient implements Client {
+                    let tokenModifiersLegend: string[] = [];
+                    for (let e in SemanticTokenModifiers) {
+                        if (isNaN(Number(e))) {
+                            tokenModifiersLegend.push(e);
+                        }
+                    }
+                    const legend: vscode.SemanticTokensLegend = new vscode.SemanticTokensLegend(tokenTypesLegend, tokenModifiersLegend);
+
+                    class SemanticTokensProvider implements vscode.DocumentSemanticTokensProvider {
+                        private client: DefaultClient;
+                        constructor(client: DefaultClient) {
+                            this.client = client;
+                        }
+
+                        public async provideDocumentSemanticTokens(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.SemanticTokens> {
+                            return new Promise<vscode.SemanticTokens>((resolve, reject) => {
+                                this.client.notifyWhenReady(() => {
+                                    let uriString: string = document.uri.toString();
+                                    let id: number = ++abortRequestId;
+                                    let params: GetSemanticTokensParams = {
+                                        id: id,
+                                        uri: uriString
+                                    };
+                                    this.client.languageClient.sendRequest(GetSemanticTokensRequest, params)
+                                        .then((tokensResult) => {
+                                            if (tokensResult.canceled) {
+                                                reject();
+                                            } else {
+                                                if (tokensResult.fileVersion !== this.client.openFileVersions.get(uriString)) {
+                                                    reject();
+                                                } else {
+                                                    let builder: vscode.SemanticTokensBuilder = new vscode.SemanticTokensBuilder(legend);
+                                                    tokensResult.tokens.forEach((token) => {
+                                                        builder.push(token.line, token.character, token.length, token.type, token.modifiers);
+                                                    });
+                                                    resolve(builder.build());
+                                                }
+                                            }
+                                        });
+                                    token.onCancellationRequested(e => this.client.abortRequest(id));
+                                });
+                            });
```

Comment:
Can this returned object be broken down into functions that can be called with `await` so it's easier to follow than the nested results?

Reply:
I'd like to do this in a subsequent change (when I move the providers into files), as this is currently a common pattern for multiple provides.